### PR TITLE
Add ``mypy`` to ``pre-commit``, and correct and export inlined type hints

### PR DESCRIPTION
**Pull request by graingert**
the type hints in the pyi files were invalid, eg using `builtins.any` instead of `typing.Any`

this PR checks the type hints using mypy with a pre-commit config adapted from dask/distributed and fixes the errors

### CODE REVIEW

File: dask_deltatable/core.py [hunk 1]
```diff
@@ -337,9 +337,9 @@ def vacuum(
     path: str,
     retention_hours: int = 168,
     dry_run: bool = True,
-    storage_options: Dict[str, str] = None,
-    delta_storage_options: Dict[str, str] = None,
-) -> None:
+    storage_options: Optional[Dict[str, str]] = None,
+    delta_storage_options: Optional[Dict[str, str]] = None,
+) -> list[str]:
```
Comment (graingert) [inline dask_deltatable/core.py:342]:
```suggestion
) -> List[str]:
```

File: dask_deltatable/core.py [hunk 2]
```diff
@@ -1,34 +1,34 @@
...
 from pyarrow import dataset as pa_ds
 
 
 class DeltaTableWrapper:
     path: str
-    version: int
-    columns: List[str]
-    datetime: str
-    storage_options: Dict[str, any]
+    version: Optional[int]
```
Comment (phofl) [inline dask_deltatable/core.py:19]: I think it would be better to use ``int | None`` (this is the new syntax)

Comment (phofl) [reply dask_deltatable/core.py:19]: Also dict instead of Dict, list instead of List, ...

**Review by phofl — COMMENTED** [review]
Couple of comments with regards to newer typing style

File: dask_deltatable/core.py (hunk 2 above)
Comment (graingert) [reply dask_deltatable/core.py:19]: This is still supporting python 3.8 without `from __future__ import annotations`

Comment (phofl) [reply dask_deltatable/core.py:19]: Any reason not to import from future? This looks good to me if there are reasons against this

Comment (graingert) [reply dask_deltatable/core.py]: I added `from __future__ import annotations` and pyupgrade fixed the calls to `List`/`Dict`/`Optional` but I needed to add flake8 to detect the unused imports

**Review by phofl — APPROVED** [review]
lgtm, I don' t have merge permissions though

**Comment by graingert** [comment]
> lgtm, I don' t have merge permissions though

I think it might be only @jrbourbeau with merge permissions here

**Review by jrbourbeau — APPROVED** [review]

**Pull request merged**